### DISCUSSION TITLE
Editing ScaledObject on Workload in Production

### DISCUSSION THREAD

**Discussion by corradomatt** [kedacore/keda#6334] [Q&A / Need Help] [answered]
Hey all!  First, thanks for a great project while I'm not using it in production yet testing is going very well and I'm excited to see it in action.

I've been running load tests using Keda ScaledObjects to control the HPA for my Deployment.  I'm using the `gcp-stackdriver` trigger on a custom GCP log metric from the workload itself and it's been working great.  However, I wanted to test the process for making adjustments to the `metadata.targetValue` of my trigger and noticed something concerning. I have the `minReplicaCount` currently set to `10` and during a load test when I adjusted the `targetValue` I noticed that my workload was scaled down from 170 pods to 10 before the new calculation kicked in a scaled it up to the expected `200+` pods.   The process happened quickly enough and my load test hardly noticed the issue, but I know this wouldn't be the case in production.

How can I ensure that the HPA doesn't scale down my workload to the `minReplicaCount` value when making minor adjustments to the ScaleObject?

**Comment by corradomatt** [comment]
I think that I see the issue in the `keda-operator` logs...

```
{
  "error": "rpc error: code = Canceled desc = grpc: the client connection is closing",
  "level": "error", "logger":"gcp_stackdriver_scaler",
  "msg":"error getting metric value",
  ...
}
```

```
{
  "error": "scaler with id 0 not found, len = 0, cache has been probably already invalidated",
  "level": "error",
  "logger": "scale_handler",
  "msg": "error getting scale decision",
  ...
}
```

It looks like the connection to GCP metrics breaks and then the scaler uses the `fallback.replicas` number.  Here's what I've set for the fallback values...

```yaml
fallback:
    failureThreshold:      3
    replicas:               10
```

Should I increase the `failureThreshold` value maybe?  Or do I just need to adjust the `fallback.replicas` to match what is currently running when I make any adjustment?

**Reply by zroubalik** [reply]
This might be the case, I would definitely try that.

**Reply by corradomatt** [reply]
I tried increasing the failureThreshold to `10` but that didn't help unfortunately.  For now, if I make changes, I update the `fallback.replicas` value to whatever the current pod count is for the workload.  That helps but I'll keep playing around with the `failureThreshold` value to see if I can find something that works.

**Reply by rickbrouwer** [reply]
I will also look later to see if I can find something. I was able to reproduce this.

**Reply by rickbrouwer** [reply]
Do you use ArgoCD by chance?

**Reply by corradomatt** [reply]
I do!  BUT in this case ArgoCD wasn't controlling the ScaledObject yet.  I was managing the ScaledObject by hand during testing and all edits were being done with `kubectl edit scaledobject X`.

**Reply by rickbrouwer** [reply]
Thanks for the explanation. I've had a look at the code and done some tests.
I think it's possible to modify the `doFallbackScaling` function in the `scale_scaledobjects.go`. That would mean a new parameter that indicates whether you want to use the currentReplicas of that moment so that it doesn't scale back but keeps the current number. It will look something like this:

```
fallback:
    failureThreshold: 3
    replicas: 10
    useCurrentReplicas: true
```
Or maybe something like useCurrentAsMinimum, so that it always uses the current number unless the replicas from the fallback are higher.

Would this be something that would help you @corradomatt ?
I don't know if this is a desirable option, otherwise I might be able to make a contribution. WDYT @zroubalik ?

**Reply by corradomatt** [reply]
That sounds like a great solution since it would be dynamic.

**Reply by rickbrouwer** [reply]
I have created the following issue: https://github.com/kedacore/keda/issues/6450.

I have basically finished the code and tested it on our own cluster. The only thing is that I still have to make tests, which will cost me a lot of time. (Yes, I know, it is wrong to first make code and then the tests, but I wanted to see it with my own eyes first ;) )

**Reply by rickbrouwer** [reply]
I have created the following PR: https://github.com/kedacore/keda/pull/6464.

**Reply by JorTurFer** [reply]
Hello
Sorry because I'm so late in the discussion :( 
I think that 
```yaml
fallback:
    failureThreshold: 3
    replicas: 10
    useCurrentReplicas: true
```
is actually almost like not setting the fallback. I mean, in the scenario that you have described (having 170 pods and fallback with 10), not having the fallback will just keep your 170 replicas untouched. This feature can be a benefit in the other direction (having 3 instances and fallback to 10). 

When KEDA applies the fallback, it shows the fact clearly in the logs, have you seen a log in KEDA operator saying that the fallback has been done? I ask because the scenario that you described (scaling in to a lower value and just instantly scaling out to the right value) usually happens because during the workload modification, the `replicas` value is set. When you modified the parameters, did you apply also the workload manifest (deployment, statefulset, etc) or just modify the ScaledObject? In case of applying all the files, does your workload manifest have the value `replicas` defined?

**Reply by rickbrouwer** [reply, marked as answer]
The proposal is later changed to a fallback behavior:

```
fallback:
    failureThreshold: 3
    replicas: 10
    behavior: <options>
```
Where the options are now:

`behavior: “useCurrentReplicasAsMinimum”`

or

`behavior: “static” `

So, it is a minimum. If the `fallback.replicas` is higher, that number will be used.
This is only useful if you want to make sure that it never scales down. So it either stays the same (current replicas) or higher (fallback replicas).

What might also be important in this discussion, but only as an addition as root cause to the initial problem.
As I see it, is that corradomatt's adjustment immediately puts it in the [doFallbackScaling](https://github.com/kedacore/keda/blob/main/pkg/scaling/executor/scale_scaledobjects.go#L169). At this piece of code, he immediately falls into a fallback and does nothing with the `failureThreshold` (that is also why increasing this value did not help him).  If KEDA **does** have to take the `failureThreshold` into account at this place, a check here on the `NumberOfFailures` might also be a quick fix.
